src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/world/GenericBlockHighlighter.java

```diff
@@ -56,34 +56,68 @@ public void onWorldRenderLast(RenderWorldLastEvent event) {
 	@SubscribeEvent
 	public void onTick(TickEvent.ClientTickEvent ev) {
 		if (ev.phase != TickEvent.Phase.END) return;
-		highlightedBlocks.removeIf(it -> !isValidHighlightSpot(it) || !canPlayerSeeBlock(it.getX(), it.getY(), it.getZ()));
+		highlightedBlocks.removeIf(it -> !isValidHighlightSpot(it) ||
+			!canPlayerSeeNearBlocks(it.getX(), it.getY(), it.getZ()));
 	}
 
 	protected boolean canPlayerSeeBlock(double xCoord, double yCoord, double zCoord) {
 		EntityPlayerSP p = Minecraft.getMinecraft().thePlayer;
 		if (p == null) return false;
-		World w = p.worldObj;
-		MovingObjectPosition hitResult = w.rayTraceBlocks(
-			new Vec3(p.posX, p.posY + p.eyeHeight, p.posZ),
-			new Vec3(xCoord, yCoord, zCoord),
-			false,
-			true,
-			true
-		);
-		BlockPos bp = new BlockPos(xCoord, yCoord, zCoord);
+		Vec3 playerPosition = new Vec3(p.posX, p.posY + p.eyeHeight, p.posZ);
+		MovingObjectPosition hitResult = rayTraceBlocks(p.worldObj, playerPosition, xCoord, yCoord, zCoord);
+		return canSee(hitResult, new BlockPos(xCoord, yCoord, zCoord));
+	}
+
+	protected boolean canPlayerSeeNearBlocks(double x, double y, double z) {
+		EntityPlayerSP p = Minecraft.getMinecraft().thePlayer;
+		if (p == null) return false;
+		World world = p.worldObj;
+		Vec3 playerPosition = new Vec3(p.posX, p.posY + p.eyeHeight, p.posZ);
+		MovingObjectPosition hitResult1 = rayTraceBlocks(world, playerPosition, x, y, z);
+		MovingObjectPosition hitResult2 = rayTraceBlocks(world, playerPosition, x + 1, y, z);
+		MovingObjectPosition hitResult3 = rayTraceBlocks(world, playerPosition, x + 1, y + 1, z);
+		MovingObjectPosition hitResult4 = rayTraceBlocks(world, playerPosition, x + 1, y + 1, z + 1);
+
+		MovingObjectPosition hitResult5 = rayTraceBlocks(world, playerPosition, x, y + 1, z + 1);
+		MovingObjectPosition hitResult6 = rayTraceBlocks(world, playerPosition, x, y + 1, z);
+		MovingObjectPosition hitResult7 = rayTraceBlocks(world, playerPosition, x + 1, y, z + 1);
+		MovingObjectPosition hitResult8 = rayTraceBlocks(world, playerPosition, x, y + 1, z);
```

Comment:
im fairly sure that you should check if the next block is nonsolid (and perhaps you can also short circuit some of these raytraces, since a raytrace is fairly expensive)

Reply:
i did the short circuit thing but couldn't figure it out a way to raytrace only visible sides using near blocks. i think this is good enough it runs pretty well for me